command/debug/debug.go

```diff
@@ -0,0 +1,691 @@
+func (c *cmd) createArchive() error {
+	f, err := os.Create(c.output + debugArchiveExtension)
+	if err != nil {
+		return fmt.Errorf("failed to create compressed archive: %s", err)
+	}
+	defer f.Close()
+
+	g := gzip.NewWriter(f)
+	defer g.Close()
+	t := tar.NewWriter(f)
+	defer t.Close()
+
+	err = filepath.Walk(c.output, func(file string, fi os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk filepath for archive: %s", err)
+		}
+
+		header, err := tar.FileInfoHeader(fi, fi.Name())
+		if err != nil {
+			return fmt.Errorf("failed to create compressed archive header: %s", err)
+		}
+
+		header.Name = filepath.Join(filepath.Base(c.output), strings.TrimPrefix(file, c.output))
+
+		if err := t.WriteHeader(header); err != nil {
+			return fmt.Errorf("failed to write compressed archive header: %s", err)
+		}
+
+		// Only copy files
+		if !fi.Mode().IsRegular() {
+			return nil
+		}
+
+		f, err := os.Open(file)
+		if err != nil {
+			return fmt.Errorf("failed to open target files for archive: %s", err)
+		}
+
+		if _, err := io.Copy(t, f); err != nil {
+			return fmt.Errorf("failed to copy files for archive: %s", err)
+		}
+
+		f.Close()
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to walk output path for archive: %s", err)
+	}
+
+	// Remove directory that has been archived
+	err = os.RemoveAll(c.output)
+	if err != nil {
+		return fmt.Errorf("failed to remove archived directory: %s", err)
+	}
+
+	return nil
+}
+
+// defaultTargets specifies the list of all targets that
+// will be captured by default
+func (c *cmd) defaultTargets() []string {
+	return append(c.dynamicTargets(), c.staticTargets()...)
+}
+
+// dynamicTargets returns all the supported targets
+// that are retrieved at the interval specified
+func (c *cmd) dynamicTargets() []string {
+	return []string{"metrics", "logs", "pprof"}
+}
+
+// staticTargets returns all the supported targets
+// that are retrieved at the start of the command execution
+func (c *cmd) staticTargets() []string {
+	return []string{"host", "agent", "cluster"}
+}
+
+func (c *cmd) Synopsis() string {
+	return synopsis
+}
+
+func (c *cmd) Help() string {
+	return c.help
+}
+
+const synopsis = "Records a debugging archive for operators"
+const help = `
+Usage: consul debug [options]
+
+  Monitors a Consul agent for the specified period of time, recording
+  information about the agent, cluster, and environment to an archive
+  written to the specified path.
+
+  If ACLs are enabled, an 'operator:read' token must be supplied in order
+  to perform this operation.
+
+  To create a debug archive in the current directory for the default
+  duration and interval, capturing all information available:
+
+      $ consul debug
+
+  Flags can be used to customize the duration and interval of the
+  operation. Duration is the total time to capture data for from the target
+  agent, whereas how often to capture dyanmic data for the length of the
```

Comment:
typo: dyanmic. I'd consider rephrasing that to be two simple sentences:

> Duration is the total time to capture data for from the target agent
> Interval controls how often to dynamic data such as metrics are scraped

Also, I totally missed this distinction on first reading.

Would it make more sense to make `duration` optional and just keep capturing until the operator kills the process? Or maybe at least default to 24 hours or something in case they actually forget and leave it running? I know if I was trying to repro a bug I wouldn't want to guess how long it might take and would rather just leave this running until I'm done? If we handled SIGINT and stopped the collection and wrote out final file that seems nicer than making them pick a duration up front.

Also does that mean that Ctrl-C on a debug process that's been running for an hour currently just looses all the output? Maybe not I didn't read closely enough but if some of this is considered already maybe mentioning that right here would be useful?